src/persist-types/src/arrow.rs

```diff
@@ -37,49 +38,112 @@ mod proto {
 }
 pub use proto::ProtoArrayData;
 
-impl RustType<ProtoArrayData> for arrow::array::ArrayData {
-    fn into_proto(&self) -> ProtoArrayData {
-        ProtoArrayData {
-            data_type: Some(self.data_type().into_proto()),
-            length: u64::cast_from(self.len()),
-            offset: u64::cast_from(self.offset()),
-            buffers: self.buffers().iter().map(|b| b.into_proto()).collect(),
-            children: self.child_data().iter().map(|c| c.into_proto()).collect(),
-            nulls: self.nulls().map(|n| n.inner().into_proto()),
+/// Extract the list of fields for our recursive datatypes.
+fn fields_for_type(data_type: &DataType) -> &[FieldRef] {
+    match data_type {
+        DataType::Struct(fields) => fields,
+        DataType::List(field) => std::slice::from_ref(field),
+        DataType::Map(field, _) => std::slice::from_ref(field),
+        _ => &[],
```

Comment:
it's a bit tedious, but do you mind explicitly listing all of the `DataType`s we currently support, and then adding an `other => unimplemented!("...")` branch? This should catch issues if we ever support other recursive types like `ListView`